test/smoke.test.ts

```diff
@@ -209,52 +265,46 @@ describe("Smoke test", () => {
         title: "My test issue",
       });
 
-      expect(HttpsProxyAgent).toHaveBeenCalled();
-
       const [call] = fetchSandbox.calls();
       expect(call[0]).toEqual(
         "https://api.github.com/repos/octocat/hello-world/issues",
       );
-      expect((call[1] as RequestOptions).agent).toBeInstanceOf(HttpsProxyAgent);
     },
   );
+  describe("customFetch", () => {
+    afterAll(() => {
+      delete process.env.HTTPS_PROXY;
+      jest.clearAllMocks();
+    });
 
-  it("Uses the explicitly provided request.agent value if it's provided", async () => {
-    process.env.GITHUB_TOKEN = "secret123";
-    process.env.GITHUB_ACTION = "test";
-    process.env.HTTPS_PROXY = "https://127.0.0.1";
-
-    const fetchSandbox = fetchMock.sandbox();
-    const mock = fetchSandbox.post(
-      "path:/repos/octocat/hello-world/issues",
-      { id: 1 },
-      {
-        body: {
-          title: "My test issue",
-        },
-      },
-    );
+    it("should call undiciFetch with the correct dispatcher", async () => {
+      process.env.HTTPS_PROXY = "https://127.0.0.1";
+      const expectedAgent = new ProxyAgent("https://127.0.0.1");
 
-    expect(Octokit).toBeInstanceOf(Function);
-    const octokit = new Octokit({
-      auth: "secret123",
-      request: {
-        fetch: mock,
-        agent: null,
-      },
-    });
-    await octokit.request("POST /repos/{owner}/{repo}/issues", {
-      owner: "octocat",
-      repo: "hello-world",
-      title: "My test issue",
-    });
+      jest.mock("../src", () => {
+        const actualModule = jest.requireActual("../src");
+        return {
+          ...actualModule,
+          getProxyAgent: jest.fn(() => expectedAgent),
+        };
+      });
+      expect(JSON.stringify(getProxyAgent())).toBe(
+        JSON.stringify(expectedAgent),
+      );
 
-    expect(HttpsProxyAgent).toHaveBeenCalled();
+      // mock undici.fetch to extract the `dispatcher` option passed in.
+      // this allows us to verify that `customFetch` correctly sets
+      // the dispatcher to `expectedAgent` when HTTPS_PROXY is set.
+      let dispatcher: any;
+      (undici.fetch as jest.Mock).mockImplementation(
+        (_url: string, options: any) => {
+          dispatcher = options.dispatcher;
 
-    const [call] = fetchSandbox.calls();
-    expect(call[0]).toEqual(
-      "https://api.github.com/repos/octocat/hello-world/issues",
-    );
-    expect((call[1] as RequestOptions).agent).toBeNull();
+          return Promise.resolve(new Response());
+        },
+      );
+      await customFetch("http://api.github.com", {});
+      expect(JSON.stringify(dispatcher)).toEqual(JSON.stringify(expectedAgent));
```

Comment:
I believe these JSON.stringify hacks are similar to the others present we've been seeing and can be cleaned up when https://github.com/octokit/core.js/issues/588 is addressed.